src/saml20_clj/sp/request.clj

```diff
@@ -33,53 +55,90 @@
-                                     :ProviderName                sp-name
-                                     :IsPassive                   false
-                                     :Destination                 idp-url
-                                     :AssertionConsumerServiceURL acs-url}
-                                    [:saml:Issuer
-                                     {:xmlns:saml "urn:oasis:names:tc:SAML:2.0:assertion"}
-                                     issuer]
-                                    ;;[:samlp:NameIDPolicy {:AllowCreate false :Format saml-format}]
-                                    ]))]
+  (let [request (coerce/->Element
+                  (coerce/->xml-string
+                    (make-auth-xml request-id instant sp-name idp-url acs-url issuer)))]
     (when state-manager
       (state/record-request! state-manager (.getAttribute request "ID")))
     (if-not credential
       request
       (or (crypto/sign request credential)
           (throw (ex-info "Failed to sign request" {:request request}))))))
 
-(defn uri-query-str
-  ^String [clean-hash]
-  (codec/form-encode clean-hash))
+(defn- add-query-params
+  "Add query parameters to a URL.
+
+  (add-query-params \"http://example.com\" {:a \"b\" :c \"d\"}
+  ;; => \"http://example.com?a=b&c=d\""
+  [url params]
+  (str url (if (str/includes? url "?") "&" "?") (codec/form-encode params)))
 
 (defn idp-redirect-response
   "Return Ring response for HTTP 302 redirect."
   [saml-request idp-url relay-state]
-  {:pre [(some? saml-request) (string? idp-url) (string? relay-state)]}
-  (let [saml-request-str (if (string? saml-request)
-                           saml-request
-                           (coerce/->xml-string saml-request))
-        url              (str idp-url
-                              (if (str/includes? idp-url "?")
-                                "&"
-                                "?")
-                              (let [saml-request-str (encode-decode/str->deflate->base64 saml-request-str)]
-                                (uri-query-str
-                                 {:SAMLRequest saml-request-str, :RelayState relay-state})))]
+  {:pre [(some? saml-request)
+         (string? idp-url)
+         (string? relay-state)]}
+  (let [saml-request-str (cond-> saml-request
+                           (not (string? saml-request)) coerce/->xml-string)
+        saml-request-str (encode-decode/str->deflate->base64 saml-request-str)
+        url              (add-query-params idp-url {:SAMLRequest saml-request-str
+                                                    :RelayState relay-state})]
     {:status  302 ; found
      :headers {"Location" url}
      :body    ""}))
+
+;; Wanted to call this make-request-xml, but it gets exported in core.clj, which
+;; warrants the request prefix
+(defn make-logout-request-xml [& {:keys [request-id instant idp-url issuer user-email]
```

Comment:
Man I need to add Kondo to this repo

Add a docstring please 🙇‍♂️

Reply:
Done